lib/github.rb

```diff
@@ -13,53 +13,76 @@ def self.retrieve_student_info(repo, cohort)
     # Catalog the list of students who have submitted
     pr_student_list = []
     pr_data.parsed_response.each do |pull_request|
       # Individual project
-      if individual
-        student_hash = individual_student(students, pull_request)
-        pr_student_list << student_hash if student_hash != nil
+      if repo.individual
+        submission = individual_student(students, repo, pull_request)
+        pr_student_list << submission if submission != nil
       else # Group project
-        pr_student_list.concat(group_project(students, pull_request))
+        pr_student_list.concat(group_project(students, pull_request, repo))
       end
     end
 
     return pr_student_list
   end
 
-  def self.add_missing_students(pr_student_list, students)
-    ids = pr_student_list.compact.map { |s| s.student_model.id }
+  def self.add_missing_students(submissions, students, repo)
+    submitted_students = submissions.map{ |submit| submit.student.id }
+    missing_students = students.reject { |stud| submitted_students.include?(stud.id) }
 
-    # Map list of students against the students who have submitted
-    missing_students = students.map { |s| !ids.include?(s.id)? s : nil }
+    missing_students.each do |stud|
+      # Do we already have a submission for this student?
+      submit = Submission.find_by(student: stud, repo: repo)
 
-    missing_students.compact.each do |missing_student_model|
-      if missing_student_model
-        pr_student_list << PRStudent.new(missing_student_model)
+      if submit
+        submissions << submit
+      else
+        submit = Submission.new(student: stud, repo: repo)
+
+        if submit.save
+          submissions << submit
+        end
       end
     end
 
-    return pr_student_list
+    return submissions
   end
 
-  def self.individual_student(students, data)
-    return create_student(students, data["user"]["login"].downcase, data["created_at"], data["html_url"])
+  def self.individual_student(students, repo, data)
+    return create_student(students, data["user"]["login"].downcase, data["created_at"], repo, data["html_url"])
   end
 
-  def self.create_student(students, user, created_at, repo_url)
+  def self.create_student(students, user, created_at, repo, pr_url)
     student_model = students.find{ |s| s.github_name == user }
+
     if student_model
-        student = PRStudent.new(student_model, user.downcase, DateTime.parse(created_at), repo_url)
-        return student
+      # Do we already have a submission for this student?
+      prev = Submission.find_by(student: student_model, repo: repo)
+
+      if prev
+        # update
+        prev.submitted_at = created_at
+        prev.pr_url = pr_url
+        if prev.save
+          return prev
+        end
+      else
+        submit = Submission.new(student: student_model, submitted_at: created_at, repo: repo, pr_url: pr_url)
+
+        if submit.save
+          return submit
+        end
```

Comment:
I think this whole section could be rewritten using `find_or_initialize_by` and `tap`:
```ruby
return Submission.find_or_initialize_by(student: student_model, repo: repo).tap |submit|
  submit.submitted_at = created_at
  submit.pr_url = pr_url
  submit.save
end
```

However it would then be necessary to:
1. Filter out PR submissions from people not in our student DB in an earlier stage
1. Change [Line 30](https://github.com/Ada-Developers-Academy/ada-prs/pull/4/files#diff-eaf4016757109d69c9ceaceb5f16b4e9R30) and [Line 111](https://github.com/Ada-Developers-Academy/ada-prs/pull/4/files#diff-eaf4016757109d69c9ceaceb5f16b4e9R111) to check for `student.persisted?` instead of nil.

Not really sure if it's worth the effort.